backend/ringv2/ringv2.go

```diff
@@ -463,25 +463,39 @@ func (r *Ring) startWatchers(ctx context.Context, ch chan Event, name string, va
 				notifyClosing(ctx, ch)
 				return
 			case response, ok := <-itemsC:
-				if err := response.Err(); err != nil {
+				err := response.Err()
+				if err != nil {
 					notifyError(ctx, ch, err)
 				}
-				if response.Canceled || !ok {
+				if !ok || response.Canceled {
 					// The watcher needs to be reinstated
-					r.startWatchers(ctx, ch, name, values, interval, cron)
-					return
+					if ctx.Err() == nil {
+						r.startWatchers(ctx, ch, name, values, interval, cron)
+						return
+					} else {
+						continue
```

Comment:
Should we use `notifyError` here and below on line 493?

Reply:
No, `ctx.Err() != nil` only means the context has been canceled. The next iteration will terminate the loop.